magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py

```diff
@@ -0,0 +1,571 @@
+"""
+This script splits the proton MC data sample into "train"
+and "test", deletes possible failed runs (only those files
+that end up with a size < 1 kB), and generates the bash
+scripts to merge the data files calling the script "merge_hdf_files.py"
+in the following order:
+
+MAGIC:
+1) Merge the subruns into runs for M1 and M2 individually.
+2) Merge the runs of M1 and M2 into M1-M2 runs.
+3) Merge all the M1-M2 runs for a given night.
+Workingdir/DL1/Observations/Merged
+
+MC:
+1) Merges all MC runs in a node and save them at
+Workingdir/DL1/MC/PARTICLE/Merged
+
+Usage:
+$ python merging_runs_and_splitting_training_samples.py (-c config.yaml)
+
+If you want to merge only the MAGIC or only the MC data,
+you can do as follows:
+
+Only MAGIC:
+$ python merging_runs_and_splitting_training_samples.py --analysis-type onlyMAGIC (-c config.yaml)
+
+Only MC:
+$ python merging_runs_and_splitting_training_samples.py --analysis-type onlyMC (-c config.yaml)
+"""
+
+import argparse
+import glob
+import logging
+import os
+from pathlib import Path
+
+import joblib
+import numpy as np
+import yaml
+from tqdm import tqdm
+
+from magicctapipe import __version__
+
+__all__ = ["cleaning", "split_train_test", "merge", "mergeMC"]
+
+logger = logging.getLogger(__name__)
+logger.addHandler(logging.StreamHandler())
+logger.setLevel(logging.INFO)
+
+
+def cleaning(list_of_nodes, cwd):
+
+    """
+    This function looks for failed runs in each node and remove them.
+
+    Parameters
+    ----------
+    list_of_nodes : array of str
+        List of nodes where the function will look for failed runs.
+    cwd : Path
+        Current working directory
+    """
+
+    for i in tqdm(range(len(list_of_nodes)), desc="Cleaning failed runs"):
+        os.chdir(list_of_nodes[i])
+        os.system('find . -type f -name "*.h5" -size -1k -delete')
```

Comment:
this is potentially dangerous, I wondered what will happen if the chdir command fails (e.g. due to wrong directory path), if it will not start deleting files in a wrong directory, but it seems not to be the case, because it throws an error and stops the script. Nevertheless, one have to be very careful with autodeletes in scripts, and also there should be some reason why those files failed - they should be caught up earlier and possibly rerun.

Reply:
ok, I also did some better fitting of the name of the files to be deleted in an earlier script so we can do it also here. About failed files, still I would like to understand what happens there, there must be always some reason why it fails and at which stage the file is broken.

Reply:
I think that the issue is related to the 'base' MCP script: we open the file with the HDF5TableWriter, then we loop on the events. If one of the events fails and make the code crash (e.g., we have some missing tables in the input file), maybe the output file is produced but not filled, and we get to a very small and 'broken' output. But I am not sure about this

Reply:
After finalizing the scripts I will try to implement a few scripts/functions to update our databases with the error codes for the different steps; this way we should be able to manually remove failed jobs before moving to the next step and we could (probably) remove these automatic deletions. As for single users, in case of broken files the merging will fail and your job accounting will signal this: so they can then remove failed files

Reply:
I removed it (commented now, then we can remove it forever) for real data (we can check failed runs by job accounting) and improved the constraint on file name for MCs (we have no job monitoring for them, and we don't need it since we want to remove them in the final version)